### ISSUE TITLE
Newline problem of Monitor's ResultsWriter in Windows environments

### ISSUE THREAD

**Issue by hsuehch** [DLR-RM/stable-baselines3#692]
https://github.com/DLR-RM/stable-baselines3/blob/e24147390d2ce3b39cafc954e079d693a1971330/stable_baselines3/common/monitor.py#L181
When running the codes in Windows environments, the logged csv files end each line with `\r\r\n`, making empty lines between data lines.
From Python 3 documentation, adding `newline=''` as an argument of the `open` function seems able to solve the problem.
https://docs.python.org/3/library/csv.html#examples

**Comment by Miffyli** [comment]
Hey, thanks for reporting this! Can you provide a minimal code example I can copy/paste and run to replicate the (unwanted) behaviour for better clarity on what goes wrong? I think I get the idea, but I might fail replicating the issue (it will be also easier to test on linux) :)

**Comment by hsuehch** [comment]
Thank you for your quick reply!
The followings are codes that can replicate the behavior, which generates a log file as attached.
The problem seems specific in Windows environments. When I ran the same codes on Ubuntu, each line ends with `\r\n` and is no problem.
```
import gym
from stable_baselines3.common.monitor import Monitor
from stable_baselines3 import DQN

env = gym.make('CartPole-v1')
env = Monitor(env, 'Train')

model = DQN('MlpPolicy', env)
model.learn(total_timesteps=1000)
``
[Train.monitor.csv](https://github.com/DLR-RM/stable-baselines3/files/7711827/Train.monitor.csv)
`

**Comment by araffin** [comment]
> The problem seems specific in Windows environments. 

But does it creates any error?
I have been using SB3 on Windows recently and I didn't experience any issue with monitor files (the plot scripts in the RL Zoo still work fine).

**Comment by Miffyli** [comment]
> But does it create any error?

Nope, but that depends on the CSV parser you use. I am not sure if CSV has formal grammar/rules for handling empty lines, but "default" Windows text editors etc will show the contents as

```
#{"t_start": 1639486796.3546653, "env_id": "CartPole-v1"}
r,l,t

23.0,23,1.257611

35.0,35,1.259605

37.0,37,1.2616

73.0,73,1.26559

...
```

This might be a headache, depending on what you want to do with the file afterwards.

**Comment by hsuehch** [comment]
> (the plot scripts in the RL Zoo still work fine)

I have not tried RL Zoo yet.

> Nope, but that depends on the CSV parser you use.

When opening the CSV file using Office Excel, empty lines make it difficult to plot.
(Need to manually replace `\r\r\n` with `\r\n`)

**Comment by Miffyli** [comment]
@araffin are you ok with a Windows-specific (or "general") fix for this consistency? I am not sure if we should encourage using the logger tools (a single run tells so little), but on the other hand it gives people more options.

**Comment by araffin** [comment]
> are you ok with a Windows-specific (or "general") fix for this consistency? 

If the fix does not change linux/mac behavior, I would be ok.
But in the meantime, we already provide helper to load the data and it should work on windows (it is not completely documented yet, see #356 ).
The code:
https://github.com/DLR-RM/stable-baselines3/blob/be86883f36eaa706e6f4a5a57d0583209471969d/stable_baselines3/common/monitor.py#L214

**Comment by hsuehch** [comment]
> @hsuehch the fix for this would be something else, probably reworking the way lines are written with hardcoded "\n" character (and instead use some function that automatically detects OS's design). Alternatively only limit it to a single "\n" character which might still work on Windows.

@Miffyli Thank you for proposing the solution. It is good for me as long as the empty lines disappear.

> yes, I would highly discourage using something else than the provided helper, as using MS Excel won't work with multiprocessing for instance or with multiple runs.

@araffin Thank you for your suggestion. I am looking forward to the RL-zoo documentation.
A minor scenario to supplement is that when a single run takes several days and some experimental settings are tried, peeking the training curve and deciding whether to continue the run are helpful. In such a case, I feel that MS Excel is more convenient, though this is just my experience.

**Comment by Miffyli** [comment]
@hsuehch We would appreciate a PR from you to fix this issue :) . The solution I described is probably the best way go. The behaviour on linux machines should remain intact.

**Comment by araffin** [comment]
> In such a case, I feel that MS Excel is more convenient, though this is just my experience.

I would disagree with that. Using tensorboard or simply `python scripts/plot_train.py -a ppo -e Cartpole -f logs/ -w 100` from the RL Zoo are two convenient options.
They also handle multi envs properly.

**Comment by hsuehch** [comment]
> @hsuehch We would appreciate a PR from you to fix this issue :) . The solution I described is probably the best way go. The behaviour on linux machines should remain intact.

@Miffyli I would like to confirm whether
> Alternatively only limit it to a single "\n" character which might still work on Windows.

means something like `self.file_handler = open(filename, "wt", newline="\n")` instead of using `newline=""`.
I will try to create a PR for this.

**Comment by Miffyli** [comment]
> means something like self.file_handler = open(filename, "wt", newline="\n") instead of using newline="".

What I meant is that all lines end with only a single "\n", so this (the quoted part) might do the trick unless rest of the code adds additional "\n" or other characters into the mix :).

**Comment by Miffyli** [comment]
Addressed in #695
